### PR TITLE
Add blockspoiler support

### PR DESCRIPTION
**Pull request by birakattack**
Added blockspoilers to snudown. Took inspiration from Stack Overflow and decided to use `>!` as the syntax to declare a blockspoiler. Some examples a blockspoiler would look like the following:

```
>! This is a blockspoiler
```

```
>! This is a blockspoiler
this is still part of the blockspoiler

This is no longer part of the blockspoiler
```

```
>>! Blockspoiler in blockquote
```

More examples and cases can be found in `test_snudown.py`. I also added some additional cases for blockquoting as it seemed like we were missing test cases for blockquotes as well. 

I also added some instructions on how to get started developing on Mac OS X in the README. 

:eyeglasses: @spladug @pwildani

### CODE REVIEW

**Comment by birakattack** [comment]
:eyeglasses: @umbrae

File: src/markdown.c [hunk 1]
```diff
@@ -1413,6 +1413,25 @@ prefix_quote(uint8_t *data, size_t size)
 	return 0;
 }
 
+static size_t
+prefix_blockspoiler(uint8_t *data, size_t size)
```
Comment (umbrae) [inline src/markdown.c]: 🔕  Just to make sure I'm understanding this right, what's happening here is the following?

This determines if we are in a block spoiler context, and if so, at what character position on this line it begins. It allows up to 3 space characters before the `>!` syntax which denotes a block spoiler, and up to 1 space character immediately following. If it does not find that, it will return 0, denoting that no block spoiler was found.

Kinda wish the surrounding code here was better documented.

Comment (birakattack) [inline src/markdown.c]: Your assessment matches my understanding of that code - should I add some comments describing that in there?

File: src/markdown.c [hunk 2]
```diff
@@ -1517,6 +1536,48 @@ parse_blockquote(struct buf *ob, struct sd_markdown *rndr, uint8_t *data, size_t
...
+	while (beg < size) {
+		for (end = beg + 1; end < size && data[end - 1] != '\n'; end++);
+
+		pre = prefix_blockspoiler(data + beg, end - beg);
+
+		if (pre)
+			beg += pre; /* skipping prefix */
+
+		/* empty line followed by non-blockspoiler line */
+		else if (is_empty(data + beg, end - beg) &&
```
Comment (umbrae) [inline src/markdown.c]: 🔕 So is this saying that if we see something like:

```
>!
Testing
```

It will ignore the blockspoiler entirely? I think that's probably pretty fine (as that's weird behavior), just trying to work it out since this doesn't check the value of `pre` at all.

File: src/markdown.c [hunk 3]
```diff
@@ -1403,7 +1403,7 @@ prefix_quote(uint8_t *data, size_t size)
 	if (i < size && data[i] == ' ') i++;
 	if (i < size && data[i] == ' ') i++;
 
-	if (i < size && data[i] == '>') {
+	if ((i < size && data[i] == '>') && (i + 1 < size && data[i+1] != '!')) {
```
Comment (umbrae) [inline src/markdown.c]: Just out of curiosity, did you consider trying to shoehorn the spoiler stuff straight into the quote work that already existed rather than creating new methods for it? Seems like it might fit, even though I am totally unsure whether that would actually be a good idea or not.

File: src/markdown.c (hunk 2 above)
Comment (birakattack) [inline src/markdown.c]: I will add that to the tests we run to demonstrate the current output value for that

File: src/markdown.c (hunk 3 above)
Comment (birakattack) [inline src/markdown.c]: I pretty much copied the blockquote logic, so blockspoilers should behave exactly as blockquotes

Comment (birakattack) [inline src/markdown.c]: The above check was to prevent spoilers being picked up as blockquotes

File: src/markdown.c [hunk 4]
```diff
@@ -1413,6 +1413,25 @@ prefix_quote(uint8_t *data, size_t size)
...
 }
 
+static size_t
+prefix_blockspoiler(uint8_t *data, size_t size)
+{
+        size_t i = 0;
+        if (i < size && data[i] == ' ') i++;
+        if (i < size && data[i] == ' ') i++;
+        if (i < size && data[i] == ' ') i++;
+        if (i < size && data[i] == '>') i++;
```
Comment (umbrae) [inline src/markdown.c]: By this method I think this makes the starting less than optional, and:

```

! this is something

```

May trigger a spoiler block. I think you need to marry this with the `&&` down below so that it will return zero if both `>` and `!` don't exist in tandem.

Comment (birakattack) [inline src/markdown.c]: oooo good catch!

File: test_snudown.py [hunk 5]
```diff
@@ -333,8 +333,56 @@
 
     '&#x;':
         '<p>&amp;#x;</p>\n',
+    '> quotey mcquoteface':
```
Comment (pwildani) [inline test_snudown.py]: Maybe add tests for empty quotes and spoilers + varieties of whitespace?

File: src/markdown.c (hunk 1 above)
Comment (umbrae) [inline src/markdown.c]: I think it couldn't hurt, it won't match the surrounding code but hell, gotta start somewhere. :)

Comment (umbrae) [inline src/markdown.c]: (The only way it could hurt I guess is if we're wrong, this is a bit of spelunking for both of us I think.)

File: src/markdown.c (hunk 3 above)
Comment (pwildani) [inline src/markdown.c]: I think this changes the behavior of > to not detect an empty block quote at the end of the text. Not necessarily a concern though.

">", size=1, i=0, i+1 =1 which is not < size
so the if expression evaluates to: `true && true && false && unevaluated` when it used to be `true && true`

File: src/markdown.c (hunk 2 above)
Comment (birakattack) [inline src/markdown.c]:
```
    '>!\nTesting':
        '<blockquote class="md-spoiler-text">\n<p>Testing</p>\n</blockquote>\n',

    '>!\n\nTesting':
        '<blockquote class="md-spoiler-text">\n</blockquote>\n\n<p>Testing</p>\n',
```

**Comment by umbrae** [comment]
💅  just to say I'm done reviewing. A healthy note of skepticism on my review as I haven't written C in about 12 years, so some of the pointer stuff is pretty murky.

Excited for this!

**Comment by spladug** [comment]
Let's chat on monday about how to do a stress test on a large corpus of real comments. Probably also want to run the fuzzer stuff Jordan added here to find potential vulnerabilities.

**Comment by andre-d** [comment]
Might want to even regression test this change against the old renderer and
just check for any output differences.

File: src/markdown.c (hunk 3 above)
Comment (spladug) [inline src/markdown.c]: also the `> ` (blank space) case seems to be unchecked for spoiler-blockquotes? should there be a `data[i+2] == ' '` equivalent of the check on 1407 for in spoilerquotes? (i'm not super clear on what that case does)

**Comment by spladug** [comment]
:nail_care: just one question. my old testing scripts appear to have disappeared so we can just make new ones i suppose :( looks good!

**Comment by birakattack** [comment]
@spladug Sounds good! Maybe we can build those scripts again next snoo's day and continue down this path. May need to lean on your expertise with regards to how, and where to grab this data.

File: src/markdown.c [hunk 6]
```diff
@@ -2331,6 +2391,9 @@ parse_block(struct buf *ob, struct sd_markdown *rndr, uint8_t *data, size_t size
 		else if (prefix_quote(txt_data, end))
 			beg += parse_blockquote(ob, rndr, txt_data, end);
 
+                else if (prefix_blockspoiler(txt_data, end))
```
Comment (GithubSecTester) [inline src/markdown.c]: Nit: this should be indented using tabs

File: src/markdown.c [hunk 7]
```diff
@@ -2331,6 +2391,9 @@ parse_block(struct buf *ob, struct sd_markdown *rndr, uint8_t *data, size_t size
...
+                        beg += parse_blockspoiler(ob, rndr, txt_data, end);
```
Comment (JordanMilne) [inline src/markdown.c]: nit: this should be indented with tabs

File: src/markdown.c [hunk 8]
```diff
@@ -1413,6 +1413,24 @@ prefix_quote(uint8_t *data, size_t size)
...
 
+static size_t
+prefix_blockspoiler(uint8_t *data, size_t size)
+{
+        size_t i = 0;
+        if (i < size && data[i] == ' ') i++;
+        if (i < size && data[i] == ' ') i++;
+        if (i < size && data[i] == ' ') i++;
+
+        if ((i < size && data[i] == '>') && (i < size && data[i + 1] == '!')) {
```
Comment (JordanMilne) [inline src/markdown.c]: Looks like `i < size && data[i + 1] == '!'` might allow an out-of-bounds read? I don't believe `data` is guaranteed to be null-terminated until `bufcstr()` is called either, so there could be minor weirdness here.

This expression could be simplified to `i + 1 < size && data[i] == '>' && data[i + 1] == '!'`.

File: README.markdown [hunk 9]
```diff
@@ -11,6 +11,17 @@ Install
 Run `setup.py install` to install the module.
 
 
+Setup for development on Mac OS X
+---------------------------------
+
+For Mac OS X:
+1. Install `afl-fuzz` via homebrew: `brew install afl-fuzz`
+2. Edit the `Makefile` located in `fuzzing/`
+    - Replace `afl-gcc` with `afl-clang`
```
Comment (JordanMilne) [inline README.markdown]: :+1: Nice, forgot how ancient the GCC toolchain was on OS X. Could we just detect OS X in the `Makefile` and automatically use `afl-clang`?

**Comment by birakattack** [comment]
This was run against a large corpus of comments and resulted in no diffs between the current release of snudown and the changes introduced by this PR. Furthermore, I also timed how long is took and the results are as follows:

*snudown without blockspoilers*
```
real    0m22.536s
user    0m10.783s
sys    0m4.342s
```

*snudown with blockspoilers*
```
real    0m18.995s
user    0m10.427s
sys    0m3.937s
```

**Comment by umbrae** [comment]
@birakattack does this need a haircut?

**Comment by birakattack** [comment]
:haircut:

**Comment by birakattack** [comment]
@JordanMilne Just hoping to get one last review here to make sure we are good to go.

**Comment by birakattack** [comment]
Added inline spoilers to this patch as well the syntax will be as follows:

blockspoiler:

> \>! This is a blockspoiler

inline spoiler:

> This is an \>!inline spoiler!<, that can be placed anywhere within a string

I will be adding additional tests as well, just thought I would put this up to get some feedback early on the inline spoiler implementation.

:haircut:

File: src/markdown.c [hunk 10]
```diff
@@ -1403,7 +1403,7 @@ prefix_quote(uint8_t *data, size_t size)
 	if (i < size && data[i] == ' ') i++;
 	if (i < size && data[i] == ' ') i++;
 
-	if ((i < size && data[i] == '>') && (i + 1 < size && data[i+1] != '!')) {
+	if ((i < size && data[i] == '>') && (i + 1 <= size && data[i+1] != '!')) {
```
Comment (spladug) [inline src/markdown.c]: Does that mean we're checking `data[i+1]` when `i+1` is out of bounds now?

File: src/markdown.c [hunk 11]
```diff
@@ -405,6 +405,9 @@ find_emph_char(uint8_t *data, size_t size, uint8_t c)
 		if (i == size)
 			return 0;
 
+		if (i - 1 > size && c == '<' && data[i] == c && data[i-1] == '!')
```
Comment (spladug) [inline src/markdown.c]: `i - 1 > size`?

**Comment by spladug** [comment]
:nail_care: for a couple of size check questions

**Comment by birakattack** [comment]
:haircut: Pushed fixes for the incorrect size checks, I have also kicked off the fuzzer again and will paste the results back here after a few days of running. In the meantime I was hoping to get this reviewed to make sure there are no other concerns that need to be addressed from a code perspective. 

Thanks!

**Comment by spladug** [comment]
:fish: lgtm (though definitely need jordan's input on this one!)

File: test_snudown.py [hunk 12]
```diff
@@ -459,3 +535,7 @@ def test_snudown():
         suite.addTest(case)
 
     return suite
+
+if __name__ == '__main__':
```
Comment (SSJohns) [inline test_snudown.py]: Did you mean to leave this in?

**Comment by zeantsoi** [comment]
@JordanMilne backwards support of inline spoilers is a [requirement for the redesign](https://reddit.atlassian.net/browse/CREATE-1018). What would be required to get final approval of this PR?

@birakattack if we were to enable _only_ inline spoilers, would disabling block level spoilers primarily be a matter of removing invocations of `blockspoiler`?

**Comment by JordanMilne** [comment]
Did we get results from the fuzzer on this?

**Comment by birakattack** [comment]
@JordanMilne Retrieved fuzzer results and there are no crashes/vulnerabilities that were found after a week of fuzzing.

**Comment by umbrae** [comment]
Woo! <3

**Comment by empyrical** [comment]
Just curious, was any of this based this off of my pull request here, @birakattack ?

https://github.com/reddit/snudown/pull/56

If so, could I get the open sourcerer badge on my reddit profile: `/u/Walter_Bishop_PhD`
